### PR TITLE
Convenience methods from Try[T] => {Future, Promise}[T]

### PR DESCRIPTION
**Pull request by retronym**
Motivation: https://github.com/retronym/http-json-example/blob/master/src/main/scala/Test.scala

Review by @viktorklang

### CODE REVIEW

File: src/library/scala/concurrent/Promise.scala [hunk 1]
```diff
@@ -139,7 +139,16 @@ object Promise {
...
    */
   def successful[T](result: T): Promise[T] = new impl.Promise.KeptPromise[T](Success(result))
-  
+
+  /** Creates an already completed Promise with the specified result or exception.
+   *
+   *  @tparam T       the type of the value in the promise
+   *  @return         the newly created `Promise` object
+   */
+  def fromTry[T](result: Try[T]): Promise[T] = result match {
```
Comment (viktorklang) [inline src/library/scala/concurrent/Promise.scala]: How about:

```
def fromTry[T](result: Try[T]): Promise[T] = new impl.Promise.KeptPromise[T](result)
```

and then change "def success" to call "fromTry(Success(result))" and "def failure" to call "fromTry(Failure(t))"

Comment (retronym) [inline src/library/scala/concurrent/Promise.scala]: Like it, pushed.

**Comment by retronym** [comment]
I don't love the fact that the name `fromTry` is somewhat unbalanced wrt `successful` / `failed`, but I couldn't think of anything better.

**Comment by viktorklang** [comment]
I also tried to come up with a better name, the only ones that made sense to me was "of" or "valueOf"

**Comment by heathermiller** [comment]
Also trying to think of a better name. So far can't think of anything.

`Future.fromTry(parseIpAddress(response))` isn't all that bad. It does sort of follow the to/from convention in the scala.concurrent and elsewhere in the stdlib. eg `fromExecutor` on `ExecutionContext`, or `fromNanos` on `Duration`,  `fromTraversables` on `ParArray`, etc.

So why not keep `fromTry`?

Funny alternative: `tried`, analogous to `failed`. heh.

**Comment by viktorklang** [comment]
Hehe `tried` is not that bad, but I agree with the fromX rationale

**Comment by retronym** [comment]
Rebased to resolve merge conflicts, which were only whitespaces.

@viktorklang @heathermiller Could I get your stamp of approval as an official el-gee-tee-em?

**Comment by viktorklang** [comment]
Just a quick thought: Isn't Try.toFuture more along the lines of the Scala STD Lib?

**Comment by retronym** [comment]
That means we can never split out `scala-concurrent.jar`. Try should have no knowledge of Future, IMO.

**Comment by viktorklang** [comment]
Then LGTM :)

**Comment by heathermiller** [comment]
Yeah, I agree with @viktorklang – `Try.toFuture` is more along the lines of how we do it everywhere else in the stdlib. I agree with this whole modularization effort, but are there concrete plans to shard scala.concurrent out into its own module? Just wondering if we should sacrifice clarity/convention in the stdlib because we want to reserve the possibility of modularizing a library in the future.

LGTM, in any event. I don't mean to be an extreme nitpicker or anything. In the above, I'm just mostly looking for clarification.

**Comment by retronym** [comment]
I think its pretty clear that the standard library should be layered; lower layers shouldn't know about higher layers. We can enforce this with module boundaries, but even before we have them, we can stick to that principle.

I could imagine a runtime where `scala.concurrent` isn't required or desired, let's say JavaScript. Or someone that uses a different approach to concurrency altogether. That platform should be still able to use `Try`.

We've failed on this account with parallel collections. The implementation of these is _not_ layered on top of the core collections, but rather entangled within them.

``` scala
trait GenTraversableLike[+A, +Repr] extends Any with GenTraversableOnce[A] with Parallelizable[A, parallel.ParIterable[A]] {
```

This exists to allow collections to polymorphically create the `combiner`:

```
% ack   parCombiner src/
src/library/scala/collection/CustomParallelizable.scala
16:  override protected[this] def parCombiner: Combiner[A, ParRepr] = throw new UnsupportedOperationException("")

src/library/scala/collection/immutable/Iterable.scala
33:  protected[this] override def parCombiner = ParIterable.newCombiner[A] // if `immutable.IterableLike` gets introduced, please move this there!

src/library/scala/collection/immutable/MapLike.scala
55:  protected[this] override def parCombiner = ParMap.newCombiner[A, B]

src/library/scala/collection/immutable/Seq.scala
36:  protected[this] override def parCombiner = ParSeq.newCombiner[A] // if `immutable.SeqLike` gets introduced, please move this there!

src/library/scala/collection/immutable/Set.scala
38:  protected override def parCombiner = ParSet.newCombiner[A] // if `immutable.SetLike` gets introduced, please move this there!

src/library/scala/collection/MapLike.scala
330:  protected[this] override def parCombiner = ParMap.newCombiner[A, B]

src/library/scala/collection/mutable/Iterable.scala
26:  protected[this] override def parCombiner = ParIterable.newCombiner[A] // if `mutable.IterableLike` gets introduced, please move this there!

src/library/scala/collection/mutable/MapLike.scala
61:  protected[this] override def parCombiner = ParMap.newCombiner[A, B]

src/library/scala/collection/mutable/SeqLike.scala
27:  protected[this] override def parCombiner = ParSeq.newCombiner[A]

src/library/scala/collection/mutable/SetLike.scala
72:  protected[this] override def parCombiner = ParSet.newCombiner[A]

src/library/scala/collection/Parallelizable.scala
41:    val cb = parCombiner
51:  protected[this] def parCombiner: Combiner[A, ParRepr]

src/library/scala/collection/SeqLike.scala
71:  protected[this] override def parCombiner = ParSeq.newCombiner[A]

src/library/scala/collection/SetLike.scala
79:  protected[this] override def parCombiner = ParSet.newCombiner[A]

src/library/scala/collection/TraversableLike.scala
106:  protected[this] def parCombiner = ParIterable.newCombiner[A]
```

And to polymorphically pick the parallel version of sequential collections:

```
ack 'def par\b' src/
src/library/scala/collection/concurrent/TrieMap.scala
778:  override def par = new ParTrieMap(this)

src/library/scala/collection/CustomParallelizable.scala
15:  override def par: ParRepr

src/library/scala/collection/immutable/HashMap.scala
107:  override def par = ParHashMap.fromTrie(this)

src/library/scala/collection/immutable/HashSet.scala
44:  override def par = ParHashSet.fromTrie(this)

src/library/scala/collection/immutable/Range.scala
51:  override def par = new ParRange(this)
342://    override def par = new ParRange(this)

src/library/scala/collection/immutable/Vector.scala
79:  override def par = new ParVector(this)

src/library/scala/collection/mutable/ArrayBuffer.scala
75:  override def par = ParArray.handoff[A](array.asInstanceOf[Array[A]], size)

src/library/scala/collection/mutable/ArrayOps.scala
69:  override def par = ParArray.handoff(repr)

src/library/scala/collection/mutable/ArraySeq.scala
59:  override def par = ParArray.handoff(array.asInstanceOf[Array[A]], length)

src/library/scala/collection/mutable/HashMap.scala
58:  override def par = new ParHashMap[A, B](hashTableContents)

src/library/scala/collection/mutable/HashSet.scala
63:  override def par = new ParHashSet(hashTableContents)

src/library/scala/collection/mutable/WrappedArray.scala
63:  override def par = ParArray.handoff(array)

src/library/scala/collection/parallel/ParIterableLike.scala
89: *     def par: Repr
246:  override def par: Repr = repr

src/library/scala/collection/Parallelizable.scala
40:  def par: ParRepr = {
```

So we can't ship parallel collections as a separate module. Which might be an interesting thing to do if we wanted to offer a new, incompatible version of them, as Alex has been working on. Or to slim down the standard library for android, JS etc.

An alternative design might trade off the open-world assumption (that needs the virtual calls to `par` and `newCombiner`) for a more layered design.

Anyway, long story short, I'd rather not paint ourselves further into a corner here! Happy to continue the discussion at the next Scala meeting. (not that this particular PR is such a big deal, but the we should establish a shared vision on how we architect the standard library)

**Comment by retronym** [comment]
Successful build: https://scala-webapps.epfl.ch/jenkins/job/pr-scala/2478/

**Comment by soc** [comment]
Sorry for being late to the party, but wouldn't it be possible to use extension methods to retain modularity while not breaking common API conventions?

Or what about adding the `to[T]` method which already exists for collections?

Imho, both would be preferable to introducing some inconsistency into the API design.
